chanzuckerberg/single-cell-data-portal
# feat(wmg): Add delete/reset cell types and genes

### Reviewers
**Functional:** 
@seve 
**Readability:** 
@MDunitz 
---

## Changes
1. Update `deleteSelectedGenesAndSelectedCellTypeIds()` event to no longer take payload, since data is in the state already
2. Add `tissueCellTypesFetched()` and `resetTissueCellTypes()` events to support delete and reset cell types
3. Add type `CellTypeMetadata`, which is string in `${id}~${tissue}~${name}` format. Also added corresponding serializer(`getAllSerializedCellTypeMetadata()`) and deserializer(`deserializeCellTypeMetadata ()`)
4. Update heat map to highlight selected cell types and genes, different text style for tissues, and reset style for tissues
5. Update `frontend/src/views/WheresMyGene/index.tsx` data flow. In general, reducer state now stores `selectedGenes` and `selectedCellTypeIds`, and `selectedCellTypes` is **derived state**, thus NOT needing to live in the store. Same thing for `tissuesWithDeletedCellTypes`, which is also a derived state that we cache via `useMemo()`
6. BUG FIX: The issue we had with React memory leak is due to debounced function not being cancelled, so we have that now for all debounced calls
7. Perf improvement: Split updating charts separately instead of in one single `useEffect()`, so we no longer trigger updates too frequently and unnecessarily -- we now have `useUpdateChart()`, useUpdateXAxisChart()` and `useUpdateYAxisChart()`
8. 

## Definition of Done (from ticket)

## QA steps (optional)
1. You can delete cell types and genes via the heat map
2. You can reset cell types by clicking its tissue
3. You should see tissue "Lung" in the row now, indicating multiple tissues support

## Known Issues
